src/storage/src/source/reclock.rs

```diff
@@ -7,931 +7,646 @@
-    pub fn new(as_of: Antichain<Timestamp>) -> Self {
+impl<FromTime, IntoTime> ReclockFollower<FromTime, IntoTime>
+where
+    FromTime: Timestamp,
+    IntoTime: Timestamp + Lattice,
+{
+    /// Constructs a new [ReclockFollower]
+    pub fn new(as_of: Antichain<IntoTime>) -> Self {
         Self {
             inner: Rc::new(RefCell::new(ReclockFollowerInner {
-                remap_trace: HashMap::new(),
-                since: as_of,
-                upper: Antichain::from_elem(Timestamp::minimum()),
-                source_upper: OffsetAntichain::new(),
+                remap_trace: Vec::new(),
+                as_of,
+                since: Antichain::from_elem(IntoTime::minimum()),
+                upper: Antichain::from_elem(IntoTime::minimum()),
+                source_since: Antichain::from_elem(FromTime::minimum()),
+                source_upper: MutableAntichain::new(),
             })),
         }
     }
 
-    /// Ensure the `ReclockFollower` has been initialized with trace
-    /// up to the given upper.
-    pub async fn ensure_initialized_to(&self, upper: AntichainRef<'_, Timestamp>) {
-        // Careful not to hold a `Ref` over an await point.
-        loop {
-            if PartialOrder::less_equal(&upper, &RefCell::borrow(&self.inner).upper.borrow()) {
-                return;
-            }
-            // Some short but non-0 amount of time
-            tokio::time::sleep(Duration::from_millis(100)).await
-        }
-    }
-
-    pub fn source_upper(&self) -> Ref<OffsetAntichain> {
-        // `borrow` overlaps with `std::borrow::Borrow` so we have to do this
-        Ref::map(RefCell::borrow(&self.inner), |inner| &inner.source_upper)
+    pub fn source_upper(&self) -> Antichain<FromTime> {
+        self.inner.borrow().source_upper.frontier().to_owned()
     }
 
     /// Pushes a new trace batch into this [`ReclockFollower`].
-    pub fn push_trace_batch(&mut self, batch: ReclockBatch) {
+    pub fn push_trace_batch(&mut self, mut batch: ReclockBatch<FromTime, IntoTime>) {
         let mut inner = self.inner.borrow_mut();
-        for (pid, ts, diff) in batch.updates {
-            let bindings = inner.remap_trace.entry(pid.clone()).or_default();
-            bindings.push((ts, diff));
-            bindings.sort_unstable();
-
-            inner.source_upper.advance(pid, diff);
-        }
+        let was_initialized = PartialOrder::less_than(&inner.as_of, &inner.upper);
+        // Ensure we only add consolidated batches to our trace
+        consolidation::consolidate_updates(&mut batch.updates);
+        inner.remap_trace.extend(batch.updates.iter().cloned());
+        inner.source_upper.update_iter(
+            batch
+                .updates
+                .into_iter()
+                .map(|(src_ts, _ts, diff)| (src_ts, diff)),
+        );
         inner.upper = batch.upper;
+        if !was_initialized && PartialOrder::less_than(&inner.as_of, &inner.upper) {
```

Comment:
No, it's for correctness. Generally we don't want to be compacting when we receive new batches, only when a user asks us by calling compact. But when we load the initial snapshot we do want to perform a single compact operation which calculates a correct since